### PR TITLE
tests: run test suite under host libc too

### PR DESCRIPTION
**Pull request by 64**
This PR adds support for running the test suite with the host `libc`, which should serve as a useful sanity check that the tests are functioning correctly. We have already found a few bugs in our implementation from this.

### CODE REVIEW

File: tests/glibc/getopt.c [hunk 1]
```diff
@@ -12,25 +17,27 @@ int main() {
...
 	int c;
 	c = getopt_long(test_argc, test_argv, shortopts, longopts, NULL);
+	debug(c);
 	assert(c == 'f');
 
 	c = getopt_long(test_argc, test_argv, shortopts, longopts, NULL);
-	assert(!strcmp(optarg, "abc"));
+	debug(c);
 	assert(c == 'b');
+	assert(!strcmp(optarg, "abc"));
```
Comment (64) [inline tests/glibc/getopt.c:40]: cc @no92, these changes were necessary to get the `getopt` test to run on `glibc` but it implies that our implementation in `mlibc` is broken.

File: tests/posix/popen.c [hunk 2]
```diff
@@ -2,7 +2,12 @@
 #include <string.h>
 #include <assert.h>
 
+#ifdef USE_HOST_LIBC
 #define TEST_FILE "popen.tmp"
+#else
+#define TEST_FILE "popen-host-libc.tmp"
```
Comment (no92) [inline tests/posix/popen.c:8]: This is the wrong way around.

File: tests/glibc/getopt.c (hunk 1 above)
Comment (no92) [reply tests/glibc/getopt.c:40]: Weird, they work for me on both mlibc and glibc. Also, I remember checking against the man pages, and our mlibc behavior should match that, IIRC.

As far as I can tell, these changes should not really change the behavior anyways other than using concatenated short options.

Also, #438 touches this test and implements missing code paths in `getopt_long`.

Comment (64) [reply tests/glibc/getopt.c:40]: With the double colon, the man page doesn't say that it should eat the next argument into optarg. What glibc version do you use?

**Comment by ArsenArsen** [comment]
is the build system code here duplicated?

**Comment by 64** [comment]
meson doesn't have functions, so yes

**Comment by ArsenArsen** [comment]
it has loops, and arrays

**Comment by 64** [comment]
> is the build system code here duplicated?

How's that for a refactor now?

I've also made the testing code use meson's `suite` feature, so that the output looks like this:
```
$ meson test abort
1/2 mlibc:host-libc+posix / abort        EXPECTEDFAIL    0.04s   killed by signal 6 SIGABRT
2/2 mlibc:posix / abort                  EXPECTEDFAIL    0.05s   killed by signal 6 SIGABRT
```

I think this is more useful for people since it automatically runs both the host libc test AND the mlibc test without any extra arguments. Plus one can now easily run all the posix tests with `meson test --suite posix`, or all the host libc tests with `meson test --suite host-libc`.

**Comment by Geertiebear** [comment]
#438 was merged, so this PR can be rebased.

**Comment by 64** [comment]
Blocked on a fix for `strtol` triggering UBSan.

File: tests/ansi/alloc.c [hunk 3]
```diff
@@ -18,6 +18,8 @@ int main() {
 	p = aligned_alloc(1, 1);
 	assert(p != NULL);
 
+	// It seems that glibc doesn't report error in these cases.
+#ifndef USE_HOST_LIBC
```
Comment (ArsenArsen) [inline tests/ansi/alloc.c:22]: perhaps this should test for glibc?

**Review by ArsenArsen — APPROVED** [review]

File: tests/ansi/alloc.c [hunk 4]
```diff
@@ -18,6 +18,8 @@ int main() {
 	p = aligned_alloc(1, 1);
 	assert(p != NULL);
 
+	// It seems that glibc doesn't report error in these cases.
```
Comment (Geertiebear) [inline tests/ansi/alloc.c:21]: We should probably submit a bug report then...

**Review by Geertiebear — APPROVED** [review]
LGTM, with one comment.

File: tests/ansi/alloc.c (hunk 4 above)
Comment (64) [reply tests/ansi/alloc.c:21]: Meh... it'd probably break some programs if they changed that behaviour. We'd still have to work around it here anyway, so we need to decide on a solution.

**Comment by Geertiebear** [comment]
Conflicts on ``tests/meson.build`` right now.

**Pull request merged**